src/theory/tools/wp/bir_htScript.sml

```diff
@@ -0,0 +1,614 @@
+    (bir_exec_block_n prog st n =
+       (ol, nstep, npc, st')
+    ) ==>
+    (st'.bst_status <> BST_AssumptionViolated)``,
+
+Induct_on `n` >| [
+  FULL_SIMP_TAC std_ss [bir_exec_block_n_def,
+                        bir_exec_steps_GEN_REWR_no_steps,
+                        valOf_BER_Ended_def],
+
+  REPEAT STRIP_TAC >>
+  ASSUME_TAC (SPECL [``prog:'a bir_program_t``,
+		     ``bl:'a bir_block_t``,
+		     ``st:bir_state_t``] bir_exec_block_exists) >>
+  FULL_SIMP_TAC std_ss [] >>
+  (* How to best swap variable names? *)
+  rename1 `st'''.bst_status = BST_AssumptionViolated` >>
+  rename1 `bir_exec_block prog bl st = (l',c',st')` >>
+  rename1 `st''.bst_status = BST_AssumptionViolated` >>
+  PAT_X_ASSUM ``!prog st bl ol nstep npc st'. blah``
+          (fn thm => ASSUME_TAC (SPECL [``prog: 'a bir_program_t``,
+                                        ``st':bir_state_t``] thm)
+          ) >>
+  IMP_RES_TAC test_lemma >>
+  IMP_RES_TAC bir_block_not_assume_never_assviol >>
+  IMP_RES_TAC bir_exec_block_n_block >>
+  PAT_X_ASSUM ``!n. blah``
+          (fn thm => ASSUME_TAC
+            (SPECL [``n:num``] thm)
+          ) >>
+  REV_FULL_SIMP_TAC std_ss [LET_DEF] >>
+  Cases_on `0 < c' /\
+            bir_state_COUNT_PC (F,(\pc. pc.bpc_index = 0)) st'`
+    >> (
+      FULL_SIMP_TAC std_ss [] >>
+      FULL_SIMP_TAC std_ss []
+   ) >> (
+    Cases_on `~bir_state_is_terminated st'` >| [
+      IMP_RES_TAC bir_exec_block_new_block_pc >>
+      FULL_SIMP_TAC std_ss [optionTheory.IS_SOME_EXISTS] >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [] >>
+      PAT_X_ASSUM ``(ol,nstep,npc,st'') = blah``
+	      (fn thm => ASSUME_TAC
+		(GSYM thm)
+	      ) >>
+      (* TODO: The below finish is ugly... *)
+      Cases_on `(bir_exec_block_n prog st' n)` >>
+      Cases_on `r` >>
+      Cases_on `r'` >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [] >>
+      IMP_RES_TAC bir_exec_block_n_REWR_TERMINATED >>
+      PAT_X_ASSUM ``!p n. blah``
+	      (fn thm => ASSUME_TAC
+		(SPECL [``prog:'a bir_program_t``, ``n:num``] thm)
+	      ) >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) []
+    ]
+  ) 
+]
+);
+(* Formulated in terms of bir_exec_to_labels: *)
+val bir_prog_not_assume_never_assviol_labels =
+  store_thm("bir_prog_not_assume_never_assviol_labels",
+  ``!ls prog st bl ol c_st c_l st'.
+    (bir_get_current_block prog st.bst_pc = SOME bl) ==>
+    bir_prog_has_no_assumes prog ==>
+    (st.bst_status <> BST_AssumptionViolated) ==>
+    (bir_exec_to_labels ls prog st =
+       BER_Ended ol c_st c_l st'
+    ) ==>
+    (st'.bst_status <> BST_AssumptionViolated)``,
+
+REPEAT STRIP_TAC >>
+subgoal `bir_exec_to_labels_n ls prog st 1 =
+           BER_Ended ol c_st c_l st'` >- (
+  FULL_SIMP_TAC (std_ss++holBACore_ss) [bir_exec_to_labels_def]
+) >>
+IMP_RES_TAC bir_exec_to_labels_n_TO_bir_exec_block_n >>
+IMP_RES_TAC bir_prog_not_assume_never_assviol
+);
+
+val bir_exec_to_labels_no_av_triple_def = Define `
```

Comment:
I find it hard to guess that `no_av` refers to the theorem in `bir_wpTheory`.

Reply:
Can you clarify?

Reply:
I said this because this definition is a copy-past from `bir_exec_to_labels_triple_def` in `bir_wpTheory` with the assumption violated disjunct removed. Maybe we should rename them:

- `bir_exec_to_labels_triple_def` => `bir_exec_to_labels_or_assumviol_triple_def`
- `bir_exec_to_labels_no_av_triple_def` => `bir_exec_to_labels_triple_def`

The motivation here is that users of the WP tool as a whole are more interested in `bir_exec_to_labels_no_av_triple_def` then in the current `bir_exec_to_labels_triple_def`.

Reply:
The reason for the current naming scheme is that `bir_exec_to_labels_triple` is mentioned much more often than `bir_exec_to_labels_no_av_triple`, which outside of its definition only exists in one theorem in HolBA. That said, I have no strong opinions on naming, so I can rename them.

Reply:
I don't totally agree to the point of interest of WP users. For compositions using assume statements for "function local" reasoning you need both. and this should be of interest for somebody who is proving a bigger program I think.

i can see arguments in both directions, for example:
- or assumviol is something more so it could be also more in the name
- the bir language itself has assume statements and therefore execution might end up in the assumviol state instead of running to the intended label

my suggestion is that we just keep it in mind for now and maybe defer the final naming to the point when wp becomes stable.

Reply:
now i see that you already implemented it. good then :)